device/cuda/src/cca/component_connection.cu

```diff
@@ -764,16 +763,49 @@ component_connection::output_type component_connection::operator()(
         }
     }
 
+    /*
+     * Allocate space on the device for input.
+     */
+    vecmem::unique_alloc_ptr<channel_id[]> channel0_device =
+        vecmem::make_unique_alloc<channel_id[]>(dmr, channel0.size());
+    vecmem::unique_alloc_ptr<channel_id[]> channel1_device =
+        vecmem::make_unique_alloc<channel_id[]>(dmr, channel1.size());
+    vecmem::unique_alloc_ptr<scalar[]> activation_device =
+        vecmem::make_unique_alloc<scalar[]>(dmr, activation.size());
+    vecmem::unique_alloc_ptr<scalar[]> time_device =
+        vecmem::make_unique_alloc<scalar[]>(dmr, time.size());
+    vecmem::unique_alloc_ptr<geometry_id[]> module_id_device =
+        vecmem::make_unique_alloc<geometry_id[]>(dmr, module_id.size());
+
+    /*
+     * Move data from host to device.
+     */
+    CUDA_ERROR_CHECK(cudaMemcpy(channel0_device.get(), channel0.data(),
+                                total_cells * sizeof(channel_id),
+                                cudaMemcpyHostToDevice));
+    CUDA_ERROR_CHECK(cudaMemcpy(channel1_device.get(), channel1.data(),
+                                total_cells * sizeof(channel_id),
+                                cudaMemcpyHostToDevice));
+    CUDA_ERROR_CHECK(cudaMemcpy(activation_device.get(), activation.data(),
+                                total_cells * sizeof(scalar),
+                                cudaMemcpyHostToDevice));
+    CUDA_ERROR_CHECK(cudaMemcpy(time_device.get(), time.data(),
+                                total_cells * sizeof(scalar),
+                                cudaMemcpyHostToDevice));
+    CUDA_ERROR_CHECK(cudaMemcpy(module_id_device.get(), module_id.data(),
+                                total_cells * sizeof(geometry_id),
+                                cudaMemcpyHostToDevice));
+
     /*
      * Store the flattened arrays in a convenience data container.
      */
     details::cell_container container;
     container.size = total_cells;
-    container.channel0 = channel0.data();
-    container.channel1 = channel1.data();
-    container.activation = activation.data();
-    container.time = time.data();
-    container.module_id = module_id.data();
+    container.channel0 = channel0_device.get();
+    container.channel1 = channel1_device.get();
+    container.activation = activation_device.get();
+    container.time = time_device.get();
+    container.module_id = module_id_device.get();
```

Comment:
I find it **very** interesting that you use a SoA data type here. Since I believe eventually we'll need to transition all our data types to an SoA layout to be compatible with ATLAS's EDM.

Out of curiosity, did you try what happens with this code when using a simple AoS layout? Does the performance suffer in such a setup? I'm genuinely very interested in this.

Reply:
To be honest this code was written so long ago I don't really remember whether I did this out of dogma or whether there was actual performance-related evidence for this... Agreed we should re-evaluate the SoA vs AoS approach on this.